### ISSUE TITLE
Improve Display Model (GDI Hooking) to support graphics/symbols and undetected text in legacy apps

### ISSUE THREAD

**Issue by Emil-18** [nvaccess/nvda#19390]
### Is your feature request related to a problem? Please describe.

The display models of other screen readers, such as JAWS and ZDSR, is able to find text NVDA's display model can't find. This makes a lot of applications that doesn't have accessibillity API support unusable with NVDA, but usable with other screen readers. Yes, the developers of those applications should add accessibillity API support, but what if they don't, or the application isn't develloped any more?

applications where NVDA's display model can't find text, but other models can includes

* Applications written with tk/tkinter (JAWS only. Completely unusable with NVDA and ZDSR)
* Pidgin (ZDSR only. Completely unusable with NVDA and JAWS)
* eclipse Java IDE (Most things works with accessibility APIs, but some text doesn't. This text is only accessible with JAWS display model)
* Mysql workbench (Works mostly with accessibility APIs. Some of the text that doesn't is accessible with both the display models in JAWS and ZDSR)

The display models of other screen readers also often support graphicks drawn to the screen. JAWS, in addition to graphics, also supports symbols, such as checkmarks, unchecked checkmarks, dropdown symbols, and so on. This can be useful if the checked/unchecked state isn't exposed through accessibility APIs, such as in the wx.CheckListBox control.
for more context, see issue #15016

### Describe the solution you'd like

Add a second display model to NVDA, that implements graphics support, support for symbols, and better/different text fetching support, so it can find the text that other display models can, but the current display model in NVDA can't. Allow the user to turn off the rendering of graphics and symbols, and allow them to switch between the new display model and the original one, in case the original works in places the new one doesn't. Allow add-ons to add their own display models. I think it is best if this is done via a path stored in the configuration, that isn't configurable by the user. If the path doesn't point to a .dll file, then use the displayModel provided by core

**Comment by tech10** [comment]
I know nothing about ZDSR, but Jaws uses video intercept, which allows for this graphics support. It also introduces a great deal more complexity, and I have encountered situations where the video intercept capability of Jaws causes problems, doesn't work, or simply increases CPU use on a system after Jaws is first launched, whether or not it continues to be used after first launch.

I can see the advantages in this feature, but the disadvantages should also be considered, the most significant of which, in my opinion, would be a component that would be far more complex than existing accessibility hooks/APIs that NVDA currently uses. I highly doubt this could be implemented in Python (I could be wrong I'm no Python developer), it's likely this would have to be written in C++ and compiled to a form Python could access. This in itself would introduce more complexity than is needed for most cases, in my opinion, and could slow things down and cause more problems rather than being more helpful.

**Comment by Emil-18** [comment]
@tech10 JAWS introduced GDI hooking  as an alternative to its graphicks driver a few years ago. NVDA also uses GDI hooking. I am not suggesting that NVDA should start to use a graphics driver, as GDI hooking is just as good anyway, but instead to improve NVDA's GDI hooking Capabilities, so it gets as good as it is in JAWS

GDI hooking is a mechanism where screen readers load a dll into all other processes on the system, and api hooks a bunch of win32 functions that draw text to the screen, making them send information about the text, font and character locations back to the screen reader, before allowing the original function to run

**Comment by cary-rowen** [comment]
I like this; it will be very helpful.

**Comment by gerald-hartig** [comment]
We would prefer to see this as an add-on as we expect this code will be very complex and challenging to maintain.

**Comment by Emil-18** [comment]
@gerald-hartig ok. That is fine for me. But for that to be as error free as possible, I think the display model code needs to be in its own separate dll, so add-ons can switch it out with their own dll without potentially causing issues to the other in process stuff

**Comment by cary-rowen** [comment]
@Emil-18 
Do you plan to work on this?

**Comment by Emil-18** [comment]
@cary-rowen I am planning on making the API more suitable for add-ons to improve the display model, since NV Access didn't wanted an improved display model in the core. I am going to add a improved display model to my Enhanced Control Support add-on

**Comment by ruifontes** [comment]
> [@cary-rowen](https://github.com/cary-rowen) I am planning on making the API more suitable for add-ons to improve the display model, since NV Access didn't wanted an improved display model in the core. I am going to add a improved display model to my Enhanced Control Support add-on

Hi! When can I find that add-on?

**Comment by Emil-18** [comment]
@ruifontes You can find it here, as of now, without any improvements to the display model. I don't know how long it is going to take until I have the display model stuff ready. https://github.com/Emil-18/enhanced-control-support

**Comment by ruifontes** [comment]
Thanks!

**Comment by Emil-18** [comment]
@gerald-hartig I did some investigations into the GDI hooking code, and it doesn't seem that hard to fix it so it gets as good as the display models of other screen readers. You don't really need to maintain it, as GDI isn't being changed anymore. If it did change, compatibility would have broken with any application using it, not just screen readers. The problems I have found so far:
* Storing memory display models by memory DCs, instead of bitmap pointers, as memory DCs draw to bitmaps, in much the same way as normal DCs draw to windows. This is what causes the display model to fail to obtain text in tkinter applications, eclypse, pigin, and others.
* Hooking FillRect, PatBlt, and BeginPaint wrong. I don't know what is wrong with FillRect, but when I removed the hook for it, some text in the iZoom application that wasn't detected previously got detected now. NVDA just clears the rectangle when PatBlt gets called, regardless of what it actualy does, (e.g inverting the screen colors). I don't think BeginPaint should be hooked at all, as it doesn't actualy do anything. I think it only notifies the window that someone is going to paint in it. Unhooking BeginnPaint makes the display model able to detect the group titles in the document formatting settings panel, such as font and document information.
* IntersectRect returns 0 if a rectangle is collapsed, but this can happen in the wild. It makes my modified display model unable to find the first letter of the button names in pigin.

I also think NVDA's display model over clears text. This can be seen in notepad ++ where it can't find the text directly under the caret when the caret is visible. In wordpad, it can't find text located on the same line as the caret. In textpad 9, when the menubar has focus, it can only detect the text on the focused menu item, as well as the one left to it
